File: 0110-balanced-binary-tree/0110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    boolean result = true;
    
    public boolean isBalanced(TreeNode root) {
        helper(root);
        return result;
    }
    
    private int helper(TreeNode node) {
        if (node == null) {
            return -1;
        }
        
        int heightL = 1 + helper(node.left);
        int heightR = 1 + helper(node.right);
        
        if (Math.abs(heightL - heightR) > 1) {
            result = false;
        }
        
        return Math.max(heightL, heightR);
    }
}

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        # approach: bottom-up DFS, compare left and right heights
        
        result = True
        
        def helper(node):
            nonlocal result
            
            if not node:
                return 0
            
            left = 1 + helper(node.left)
            right = 1 + helper(node.right)
            
            if abs(left - right) > 1:
                result = False
                
            return max(left, right)
        
        helper(root)
        return result